File: app/src/main/java/com/rubenwardy/minetestmodmanager/models/Game.java
package com.rubenwardy.minetestmodmanager.models;

import android.support.annotation.NonNull;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Game {
    public final String name;
    private final File file;

    private final Map<String, ModList> lists = new HashMap<>();

    public Game(@NonNull String name, @NonNull File file) {
        this.name = name;

        this.file = file;
    }

    public String getPath() {
        return file.getAbsolutePath();
    }

    public ModList getList(String path) {
        return lists.get(new File(path).getAbsolutePath());
    }

    public Mod getMod(String name, String author) {
        for (ModList list : lists.values()) {
            Mod mod = list.get(name, author);
            if (mod != null) {
                return mod;
            }
        }

        return null;
    }

    public boolean isValid() {
        return file.isDirectory() && new File(file, "mods").isDirectory();
    }

    public boolean isLoaded() {
        return new File(file, "games").isDirectory();
    }

    public boolean hasWorld() {
        return new File(file, "worlds/world").isDirectory();
    }

    @SuppressWarnings("ResultOfMethodCallIgnored")
    public void forceCreate() {
        new File(file, "mods").mkdirs();
        new File(file, "worlds").mkdirs();
    }

    public List<ModDir> getModPaths() {
        List<ModDir> paths = new ArrayList<>();
        paths.add(new ModDir(new File(file, "mods").getAbsolutePath(), ModList.ModListType.EMLT_MODS));
        paths.add(new ModDir(new File(file, "games/minetest_game/mods").getAbsolutePath(), ModList.ModListType.EMLT_GAME_MODS));
        return paths;
    }

    public List<ModList> getAllModLists() {
        List<ModList> ret = new ArrayList<>();
        ret.addAll(lists.values());
        return ret;
    }

    public Map<String, Mod> getAllMods() {
        Map<String, Mod> map = new HashMap<>();

        for (ModList list : lists.values()) {
            for (Mod mod : list.mods) {
                map.put(mod.name, mod);
            }
        }

        return map;
    }

    public void addList(String path, ModList list) {
        lists.put(path, list);
    }

    public boolean hasPath(String path) {
        String root;
        try {
            root = file.getCanonicalPath();
        } catch (IOException e) {
            root = file.getAbsolutePath();
        }

        File file2 = new File(path);
        try {
            path = file2.getCanonicalPath();
        } catch (IOException e) {
            path = file2.getAbsolutePath();
        }

        return path.startsWith(root);
    }

    public class ModDir {
        public String path;
        public ModList.ModListType type;

        ModDir(String absolutePath, ModList.ModListType emltGameMods) {
            this.path = absolutePath;
            this.type = emltGameMods;
        }
    }
}

File: app/src/main/java/com/rubenwardy/minetestmodmanager/models/Game.kt
package com.rubenwardy.minetestmodmanager.models

import java.io.File
import java.io.IOException
import java.util.ArrayList
import java.util.HashMap

import io.realm.RealmObject

open class Game(val name: String, private val file: File) {

    private val lists = HashMap<String, ModList>()

    val path: String
        get() = file.absolutePath

    val doesModDirExist: Boolean
        get() = file.isDirectory && File(file, "mods").isDirectory

    val gamesDirExists: Boolean
        get() = File(file, "games").isDirectory

    val hasWorld: Boolean
        get() = File(file, "worlds/world").isDirectory

    fun getListFromPath(path: String): ModList? {
        return lists[File(path).absolutePath]
    }

    fun getModFromNameAuthor(name: String, author: String): Mod? {
        for (list in lists.values) {
            val mod = list.get(name, author)
            if (mod != null) {
                return mod
            }
        }

        return null
    }

    fun forceCreate() {
        File(file, "mods").mkdirs()
        File(file, "worlds").mkdirs()
    }

    val modPaths: List<ModDir>
        get() {
            val paths = ArrayList<ModDir>()
            paths.add(ModDir(File(file, "mods").absolutePath, ModList.ModListType.EMLT_MODS))
            paths.add(ModDir(File(file, "games/minetest_game/mods").absolutePath, ModList.ModListType.EMLT_GAME_MODS))
            return paths
        }

    val allModLists: List<ModList>
        get() {
            val ret = ArrayList<ModList>()
            ret.addAll(lists.values)
            return ret
        }

    val allMods: Map<String, Mod>
        get() {
            val map = HashMap<String, Mod>()

            for (list in lists.values) {
                for (mod in list.mods) {
                    map.put(mod.name, mod)
                }
            }

            return map
        }

    fun addList(path: String, list: ModList) {
        lists.put(path, list)
    }

    fun hasPath(path: String): Boolean {
        var testPath = path
        var root: String
        try {
            root = file.canonicalPath
        } catch (e: IOException) {
            root = file.absolutePath
        }

        val file2 = File(testPath)
        try {
            testPath = file2.canonicalPath
        } catch (e: IOException) {
            testPath = file2.absolutePath
        }

        return testPath.startsWith(root)
    }

    inner class ModDir internal constructor(var path: String, var type: ModList.ModListType)

    companion object {
        fun isMTGMod(mod_name: String): Boolean {
            val mods = arrayOf("beds", "boats", "bones", "bucket", "carts", "creative", "default", "doors", "dye", "farming", "fire", "flowers", "give_initial_stuff", "killme", "screwdriver", "sethome", "sfinv", "stairs", "tnt", "vessels", "walls", "wool", "xpanes")

            for (mod in mods) {
                if (mod == mod_name) {
                    return true
                }
            }

            return false
        }
    }
}
